storage/src/main/java/io/delta/storage/HDFSLogStore.java

```diff
@@ -137,6 +139,37 @@ private void writeInternal(
                 fc.delete(tempPath, false); // recursive=false
             }
         }
+
+        msyncIfSupported(path, hadoopConf);
+    }
+
+    /**
+     * Normally when using HDFS with an Observer NameNode setup, there would be read after write
+     * consistency within a single process, so the write would be guaranteed to be visible on the
+     * next read. However, since we are using the FileContext API for writing (for atomic rename),
+     * and the FileSystem API for reading (for more compatibility with various file systems), we
+     * are essentially using two separate clients that are not guaranteed to be kept in sync.
+     * Therefore we "msync" the FileSystem instance, which is cached across all uses of the same
+     * protocol/host combination, to make sure the next read through the HDFSLogStore can see this
+     * write.
+     * Any underlying FileSystem that is not the DistributedFileSystem will simply throw an
+     * UnsupportedOperationException, which can be ignored. Additionally, if an older version of
+     * Hadoop is being used that does not include msync, a NoSuchMethodError will be thrown while
+     * looking up the method, which can also be safely ignored.
+     */
+    private void msyncIfSupported(Path path, Configuration hadoopConf) throws IOException {
+        try {
+            FileSystem fs = path.getFileSystem(hadoopConf);
+            Method msync = fs.getClass().getMethod("msync");
+            msync.invoke(fs);
+        } catch (NoSuchMethodException e) {
+        } catch (IllegalAccessException e) {
+        } catch (InvocationTargetException e) {
+        } catch (Throwable e) {
+            if (!LogStoreErrors.isNonFatal(e)) {
```

Comment:
nit: it's better to add a comment such as `We ignore non fatal errors as calling msync is best effort.`

Reply:
Same as above, Java doesn't seem to recognize that isNonFatal is catching all those cases and still complains that the exception must be declared to be thrown

Reply:
How about this
```
    private void msyncIfSupported(Path path, Configuration hadoopConf) throws IOException {
        try {
            FileSystem fs = path.getFileSystem(hadoopConf);
            Method msync = fs.getClass().getMethod("msync");
            msync.invoke(fs);
        } catch (InterruptedIOException e) {
            throw e;
        } catch (Throwable e) {
            if (e instanceof InterruptedException) {
                // Propagate the interrupt status
                Thread.currentThread().interrupt();
            }
            // We ignore non fatal errors as calling msync is best effort.
        }
    }
```
We don't want to ignore interrupt status as it's a signal to stop running. The rest should be okay to ignore.

Reply:
Java seems happy so works for me. Updated scala to just ignore NonFatal